stumpy/core.py

```diff
@@ -2559,3 +2559,76 @@ def _select_P_ABBA_value(P_ABBA, k, custom_func=None):
             MPdist = partition[k]
 
     return MPdist
+
+
+@njit(parallel=True)
+def _merge_topk_PI(PA, PB, IA, IB):
+    """
+    Merge two top-k matrix profiles PA and PB, and update PA (in place) while
+    always choosing values of PA over values of PB in case of ties. Also, update
+    IA accordingly.
+
+    Parameters
+    ----------
+    PA : numpy.ndarray
+        a (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PB.
+
+    PB : numpy.ndarray
+        a (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PA.
+
+    IA : numpy.ndarray
+        a (top-k) matrix profile indices, corresponding to PA
+
+    IB : numpy.ndarray
+        a (top-k) matrix profile indices, corresponding to PB
+
+    Returns
+    -------
+    None
+    """
+    for i in prange(PB.shape[0]):
+        start = 0
+        stop = np.searchsorted(PA[i], PB[i, -1], side="right")
+
+        for j in range(PB.shape[1]):
+            if PB[i, j] < PA[i, -1]:
+                idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start
+
+                for g in range(PB.shape[1] - 1, idx, -1):
```

Comment:
I think you've unlocked another level in your programming journey! By creating `_shift_insert_at_index`, you've not only made it more obvious what is happening but you've also removed/simplified a few lines of code.  We should look for more opportunities to refactor our codebase like this!

Btw, should lines 2600-2604 be removed too?

Reply:
> I think you've unlocked another level in your programming journey!

I hope that is true :)

> We should look for more opportunities to refactor our codebase like this!

I will let you know if I find such opportunities elsewhere

> Btw, should lines 2600-2604 be removed too?

Yes. I didn't know how to put a comment that shows that part. So, I put the comment at the beginning of `for-loop`. Maybe I should have inserted my comment just at line 2604 so it would be more clear.